core/src/main/java/com/sequenceiq/cloudbreak/service/stack/InstanceMetaDataService.java

```diff
@@ -114,7 +117,8 @@ public StackDtoDelegate saveInstanceAndGetUpdatedStack(StackDtoDelegate stack, M
         Map<String, String> subnetAzPairs = multiAzCalculatorService.prepareSubnetAzMap(environment);
         String stackSubnetId = getStackSubnetIdIfExists(stack);
         String stackAz = stackSubnetId == null ? null : subnetAzPairs.get(stackSubnetId);
-        long privateId = getFirstValidPrivateId(stack.getInstanceGroupDtos());
+        List<InstanceGroupDto> allInstanceMetadataByInstanceGroup = stackDtoService.getInstanceMetadataByInstanceGroup(stack.getId());
```

Comment:
you can create a query directly to fetch the next valud private id. I think it won't cause unnecessary traffic

Reply:
CB-17939 is created to optimize getFirstValidPrivateId